panther_hardware_interfaces/src/roboteq_driver.cpp

```diff
@@ -100,110 +98,81 @@ bool RoboteqDriver::WaitForBoot()
     rpdo_mapped[RoboteqCANObjects::position_2.id][RoboteqCANObjects::position_2.subid];
 
-  fb.motor_1.vel =
+  states.motor_1.vel =
     rpdo_mapped[RoboteqCANObjects::velocity_1.id][RoboteqCANObjects::velocity_1.subid];
-  fb.motor_2.vel =
+  states.motor_2.vel =
     rpdo_mapped[RoboteqCANObjects::velocity_2.id][RoboteqCANObjects::velocity_2.subid];
 
-  fb.motor_1.current =
+  states.motor_1.current =
     rpdo_mapped[RoboteqCANObjects::current_1.id][RoboteqCANObjects::current_1.subid];
-  fb.motor_2.current =
+  states.motor_2.current =
     rpdo_mapped[RoboteqCANObjects::current_2.id][RoboteqCANObjects::current_2.subid];
 
-  fb.fault_flags = GetByte(
-    static_cast<int32_t>(rpdo_mapped[RoboteqCANObjects::fault_script_flags.id]
-                                    [RoboteqCANObjects::fault_script_flags.subid]),
-    0);
-  fb.script_flags = GetByte(
-    static_cast<int32_t>(rpdo_mapped[RoboteqCANObjects::fault_script_flags.id]
-                                    [RoboteqCANObjects::fault_script_flags.subid]),
-    2);
+  std::unique_lock<std::mutex> lck_p(position_timestamp_mtx_);
+  states.pos_timestamp = last_position_timestamp_;
 
-  fb.runtime_stat_flag_motor_1 = GetByte(
-    static_cast<int32_t>(
-      rpdo_mapped[RoboteqCANObjects::motor_flags.id][RoboteqCANObjects::motor_flags.subid]),
-    0);
-  fb.runtime_stat_flag_motor_2 = GetByte(
-    static_cast<int32_t>(
-      rpdo_mapped[RoboteqCANObjects::motor_flags.id][RoboteqCANObjects::motor_flags.subid]),
-    1);
-
-  std::unique_lock<std::mutex> lck(rpdo_timestamp_mtx_);
-  fb.timestamp = last_rpdo_write_timestamp_;
+  std::unique_lock<std::mutex> lck_sc(speed_current_timestamp_mtx_);
+  states.vel_current_timestamp = last_speed_current_timestamp_;
 
-  return fb;
+  return states;
 }
 
-// todo check what happens when publishing is stopped (on hold - waiting for decision on changing to
-// PDO)
-void RoboteqDriver::SendRoboteqCmdChannel1(const std::int32_t cmd)
+RoboteqDriverState RoboteqDriver::ReadRoboteqDriverState()
 {
-  try {
-    SyncSDOWrite<std::int32_t>(RoboteqCANObjects::cmd_1.id, RoboteqCANObjects::cmd_1.subid, cmd);
-  } catch (const std::runtime_error & e) {
-    throw std::runtime_error(
-      "Error when trying to send channel 1 Roboteq command: " + std::string(e.what()));
-  }
+  RoboteqDriverState state;
+
+  state.fault_flags = GetByte(
+    static_cast<std::int32_t>(
+      rpdo_mapped[RoboteqCANObjects::flags.id][RoboteqCANObjects::flags.subid]),
+    0);
+  state.runtime_stat_flag_motor_1 = GetByte(
+    static_cast<std::int32_t>(
+      rpdo_mapped[RoboteqCANObjects::flags.id][RoboteqCANObjects::flags.subid]),
+    1);
+  state.runtime_stat_flag_motor_2 = GetByte(
+    static_cast<std::int32_t>(
+      rpdo_mapped[RoboteqCANObjects::flags.id][RoboteqCANObjects::flags.subid]),
+    2);
+  state.script_flags = GetByte(
+    static_cast<std::int32_t>(
+      rpdo_mapped[RoboteqCANObjects::flags.id][RoboteqCANObjects::flags.subid]),
+    3);
```

Comment:
I would prefer not to do it, certainly it won't fit with the `CANopenObject`. Instead we could move this conversion to roboteq data converters, here we would save the whole `int32_t` value to the flags variable in the `RoboteqDriverState`. I think that it will make more sense.

Reply:
I guess we can keep it this way

Reply:
okay, let's wait for @pkowalsk1 opinion

Reply:
In my opinion, using 0, 1, 2, 3 is okay. However, wouldn't it be better to retrieve the flag value only once at the beginning and store it in a variable when the idx and subidx are the same?

Reply:
that's good point, I'll change that